### ISSUE TITLE
Make sure services have been started before executing CMD

### ISSUE THREAD

**Issue by janeczku** [just-containers/s6-overlay#114]
It looks like there is currently no guarantee that all user defined services are actually started before the CMD is executed so we might have a race condition when the CMD depends on a service to be running.
Could we leverage `s6-svwait` to make sure that services have been started before running the CMD?

In https://github.com/just-containers/s6-overlay/blob/master/builder/overlay-rootfs/etc/s6/init/init-stage2#L136 insert something like this:

```BASH
        if
        {
          pipeline { s6-ls -0 -- /var/run/s6/etc/services.d }
          forstdin -0 -p -- i
          importas -u i i
          if { s6-svwait -t 1000 -u /var/run/s6/services/${i} }
        }
        if { s6-echo -- "[services.d] done." }

```

**Comment by glerchundi** [comment]
Seems reasonable, maybe this should be parameterized as well to keep backward compatibility. I'll put this on hold until I found a free slot.

Thanks

**Comment by glerchundi** [comment]
@janeczku after thinking a little bit about this, my opinion is that isn't worth it. why is `s6-supervise` trigger enough to be considered as up? Imagine a container image which has both client and server binaries, and although the default behaviour is to put the server up and running, it would be possible to launch a `CMD` against the recently created server. But, this really means that the server is really up? You'll find yourself in a race condition between supervised services considering it up while the server is not in a running state at all.

This is the eternal debate of readiness and the requirements for the people probably are completely different from yours and I don't think messing the init with multiple flags just to cover most of them is the way to go.

Perhaps an utility which does what you propose and just forwards svwait flags would be interesting. Something like `with-svwaitall flags`. So the run command would look like this:

`docker run -ti myimage with-svwaitall -U -- bash`

WDYT?

**Comment by janeczku** [comment]
> You'll find yourself in a race condition between supervised services

Race is inherit in every service init/supervision system (upstart, sysvinit). Yet, all those systems provide some way to control the *start* order (e.g. by using `Required-start:` in an init script). This eliminates the most important race: **Process creation**.

Since the s6 overlay init script processes services first ("[services.d] done." console message), users can reasonably expect that the services are in fact started before the CMD. 
`sv6wait -u/-U` guarantees just that.

> But, this really means that the server is really up?

If users make use of s6's [startup notification](http://skarnet.org/software/s6/notifywhenup.html) mechanism in their services, yes.

I don't necessarily share the view that adding s6wait to the init script is merely a means to support edge cases.
While guaranteeing service ready-ness might be an edge case, guaranteeing that the CMD process is started last is not.

**Comment by glerchundi** [comment]
@janeczku PTAL to #133 I would like to discuss the implementation before proceeding.

**Comment by glerchundi** [comment]
@janeczku i moved the PR to #160 can you take a look ASAP?

**Comment by glerchundi** [comment]
implemented in #160.
